### PR TITLE
Iss182 catchall logger

### LINKED ISSUE #182: Generic Error handling
## Objective
Similar to the way we handle `OperationalError` in the flask_api, we should have a catchall to log what happens for uncaught exeptions

## Details
(as of now)
Flask_api has the following:
```
@app.errorhandler(OperationalError)
def handle_database_error(error):
    global db
    if db is None:
        # reinit the database
        init_nimbus_db()
    else:
        # we *probably* have a bad session - try and roll it back,
        # then create a new database connection.
        db.session.rollback()
        db.session.close()
        db = None
        init_nimbus_db()
```
to catch OperationalErrors.  We should modify this/change this to handle more than just OperationalErrors (like IndexErrors, etc) and log them.

**Additional context**
Python errors all follow object oriented design (I think), so I think we can catch a generic `Exception` and check what type it is to handle different ones differently.

[1]: https://github.com/calpoly-csai/api/milestones

### PR DESCRIPTION
**Pull request by ethanahlquist**
## What's New?

I think this is good, but review would be greatly appreciated

Branch includes changes to allow generic error handling, presumably for all cases. In the SQL database, questions and their stacktraces are stored in the `ErrorLog` table. This error catching much more elegant. Caught stacktraces are still printed into stdout also. 

Changes include:
- Added `handle_all_errors()` function in flask_api.py
    - Catches all exceptions
    - Maintains previous functionality in `OperationalError`
    - Calls function to log error to database, in table (ErrorLog)

- Adding `log_error()` function in flask_api.py
    - Logs stacktrace to database.
    - takes optional question field, as to provide more error context

- Added new Entity file called ErrorLog
    - ErrorLog fields include (id, question, stacktrace, timestamp)
    - added necessary fields to Enity/ExcpectedKeys.py

- Added `format_error_log()` to database_wrapper.py
    - allows ErrorLog, to use timestamp without raising a BadDictionaryKeyError

- Added necessary imports to files

## Type of change (pick-one)
- [ ] Bug fix (_non-breaking change which fixes an issue_)
- [x] New feature (_non-breaking change which adds functionality_)
- [ ] Breaking change (_fix or feature that would cause existing functionality to not work as expected_)
- [ ] This change requires a documentation update

## How Has This Been Tested?

I have seen correct ErrorLogs in the SQL database, after running:

curl --location --request POST 'localhost:8080/ask' --header 'Content-Type: application/json' --data-raw '{ "question": "When CPE 233?" }'

Which caused a KeyError in `nimbus.answer_question(question)`
- Error was used to display the new code logging errors properly
- _After using this exploit, I fixed the error by catching it's exception_

I have also tested with functioning requests

## Checklist (check-all-before-merge)
_formatting help: `- [x]` means "checked' and `- [ ]` means "unchecked"_

- [ ] I documented my code according to the [Google Python Style Guide][1]

- [ ] I ran `./build_docs.sh` and the docs look fine

- [ ] I ran `./type_check.sh` and got no errors

- [ ] I ran `./format.sh` because it automatically cleans my code for me 😄 

- [ ] I ran `./lint.sh` to check for what "format" missed

- [ ] I added my tests to the `/tests` directory

- [ ] I ran `./run_tests.sh` and all the tests pass

[1]: https://google.github.io/styleguide/pyguide.html

### CODE REVIEW

File: flask_api.py [hunk 1]
```diff
@@ -127,31 +147,41 @@ def handle_question():
...
+            response["session"] = request_body["session"]
+        else:
+            response["session"] = generate_session_token()
+
+        return jsonify(response), SUCCESS
+
+    except Exception as e:
+        log_error(e, question)
+        response = {"answer": "oops, something went wrong... Try another question"}
+        return jsonify(response), SERVER_ERROR # Should this really be SERVER_ERROR?
```
Comment (snekiam) [inline flask_api.py:184]: Theoretically, we should determine if it is a SERVER_ERROR or a CLIENT_ERROR.  I'm not sure that's in scope for this - we used to simply 500 and return nothing useful, so this is an improvement.

**Review by snekiam — APPROVED** [review]
Looks good to me. Awesome job on this PR!

File: Entity/ErrorLog.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+from sqlalchemy import Text, Integer, Column, TIMESTAMP
+from sqlalchemy.ext.declarative import declarative_base
+
+Base = declarative_base()
+
+
+class ErrorLog(Base):
```
Comment (Jason-Ku) [inline Entity/ErrorLog.py:7]: Having a table for errors is a really great idea!

File: flask_api.py (hunk 1 above)
Comment (Jason-Ku) [reply flask_api.py:184]: One way I think you could do that is to have two top level error "interfaces" or parent modules/classes and have other errors fall into one of those categories but I agree that the changes necessary to make that happen would fall outside of the scope of this code review

**Review by Jason-Ku — APPROVED** [review]

**Pull request merged**